src/test/java/io/jenkins/plugins/analysis/core/history/PreviousRunReferenceTest.java

```diff
@@ -0,0 +1,522 @@
+     */
+    @Test
+    void twoTrueAction_withAnyResult_shouldReturnTwo() {
+        Run baseline = mock(Run.class);
+        Run previousRunMock = mock(Run.class);
+        when(baseline.getPreviousBuild()).thenReturn(previousRunMock);
+        when(previousRunMock.getResult()).thenReturn(Result.SUCCESS, Result.SUCCESS);
+
+        ResultAction action = mock(ResultAction.class);
+        when(action.isSuccessful()).thenReturn(true);
+
+        AnalysisResult analysis = mock(AnalysisResult.class);
+        when(action.getResult()).thenReturn(analysis);
+        when(analysis.getOverallResult()).thenReturn(Result.SUCCESS, Result.SUCCESS);
+
+        ResultSelector selector = mock(ResultSelector.class);
+        when(selector.get(previousRunMock)).thenReturn(Optional.of(action));
+
+        List<ResultAction> actions = Collections.singletonList(action);
+        when(previousRunMock.getActions(ResultAction.class)).thenReturn(actions);
+
+        PreviousRunReference previousRun = new PreviousRunReference(baseline, selector,
+                false);
+
+        Optional<ResultAction> resultAction1 = previousRun.getReferenceAction();
+        Optional<ResultAction> resultAction2 = previousRun.getReferenceAction();
+
+        verify(selector, times(4)).get(previousRunMock);
+        assertThat(resultAction1).isEqualTo(Optional.of(action));
+        assertThat(resultAction2).isEqualTo(Optional.of(action));
+    }
+
+    /**
+     * Test that there are two previous action, with overallResultMusBeSuccess = false.
+     */
+    @Test
+    void twoAction_withAnyResult_shouldReturnTwo() {
+        Run baseline = mock(Run.class);
+        Run previousRunMock = mock(Run.class);
+        when(baseline.getPreviousBuild()).thenReturn(previousRunMock);
+        when(previousRunMock.getResult()).thenReturn(Result.FAILURE, Result.SUCCESS);
+
+        ResultAction action = mock(ResultAction.class);
+        when(action.isSuccessful()).thenReturn(true);
+
+        AnalysisResult analysis = mock(AnalysisResult.class);
+        when(action.getResult()).thenReturn(analysis);
+        when(analysis.getOverallResult()).thenReturn(Result.FAILURE, Result.SUCCESS);
+
+        ResultSelector selector = mock(ResultSelector.class);
+        when(selector.get(previousRunMock)).thenReturn(Optional.of(action));
+
+        List<ResultAction> actions = Collections.singletonList(action);
+        when(previousRunMock.getActions(ResultAction.class)).thenReturn(actions);
+
+        PreviousRunReference previousRun = new PreviousRunReference(baseline, selector,
+                false);
+
+        Optional<ResultAction> resultAction1 = previousRun.getReferenceAction();
+        Optional<ResultAction> resultAction2 = previousRun.getReferenceAction();
+
+        verify(selector, times(4)).get(previousRunMock);
+        assertThat(resultAction1).isEqualTo(Optional.of(action));
+        assertThat(resultAction2).isEqualTo(Optional.of(action));
```

Comment:
should be changed to:

when(previousRunMock.getResult()).thenReturn(Result.FAILURE);
when(analysis.getOverallResult()).thenReturn(Result.FAILURE)
assertThat(previousRun.getReferenceAction()).contains(action);

when(previousRunMock.getResult()).thenReturn(Result.SUCCESS);
when(analysis.getOverallResult()).thenReturn(Result.SUCCESS)
assertThat(previousRun.getReferenceAction()).contains(action);

so it is easier for the reader of the code to see what exactly the difference between the two calls is.